Morphtargets

lemirep opened the pull request:
- Mesh and Animation GLTF2 parsers updated to retrieve MorphTarget animation
- Actual animation and generation of compatible shader still needs to be done

lemirep committed:
MeshParser: parse morph targets

Task-Id: KUE-639

lemirep committed:
AnimationParser: fix coding style

lemirep committed:
AnimationParser: handle the weights channel for morph targets

Task-Id: KUE-640

lemirep committed:
Accessor: initialize bufferViewIndex to -1

As 0 could be a valid index, if the bufferViewIndex isn't
initialized this will now print a warning instead of silently
using the first bufferView.

lemirep committed:
MeshParser: add morph target attributes to the geometry

lemirep committed:
MeshParser: improve warning output for Draco parsing

lemirep committed:
MeshParser::geometryAttributesDracoFromJSON tidies

- Stop using const ref &everywhere
- Move accessor access closer to where it is used

lemirep committed:
Add MorphController class

Task-Id: KUE-644

lemirep committed:
NodeParser: handle the optional weights entry

Which specify defaults weights for morph targets

lemirep committed:
GLTF2Parser: instantiate MorphController for Nodes with morph targets

lemirep committed:
GLTF2Parser: create animation channel mapping for morph targets

jalbamon reviewed (commented):
Documentation looks in order

jalbamon commented on src/core/gltf2importer/animationparser.cpp:
```diff
@@ -277,6 +282,22 @@ Qt3DAnimation::QChannel AnimationParser::animationChannelDataFromBuffer(const QS
+    // but the channel should actually expose one scalar weight for each morph targets
+    // -> one channel component for each morph target in the mesh
+    // 1 - n possibly
+    const bool isMorphTargetWeightChannel = (channelName == QStringLiteral("weights"));
```
I'm a little confused by the last 3 lines of the comment, but that could be because I have no idea about morph targets.  Just check to make sure it makes sense.

lemirep replied:
Right, I found it confusing it as well. Reworded it.

lemirep committed:
MetallicRoughnessMaterial: handle Morph Targets

- Use graphs to generate the metallicroughness vertex shader
- Only two layers for now: skinning and no-skinning
- For now we still rely on hardcoded glsl prototypes for
  morph target handling

Requires https://codereview.qt-project.org/#/c/257748/
Task-Id: KUE-641

lemirep committed:
MaterialParser: simplify code a bit

lemirep committed:
MetallicRoughnessMaterial: take in MorphController as shader parameter

It holds the weights to be used for morph target animations

lemirep committed:
MeshParser: allow attribute size of 3 or 4 for tangents

lemirep committed:
Fix substitution for discard_if and if shader nodes

Required to work properly with recent QShaderGenerator changes

lemirep commented:
Updated with latest dev. Substitution for the if node had to be fixed for the shader generation to work correctly and not break multi UV support. Tested along with https://codereview.qt-project.org/#/c/258754/

DaveeFTW commented on src/core/morphcontroller.cpp:
```diff
@@ -0,0 +1,393 @@
+    weights.reserve(m_count);
+
+    for (int i = 0; i < m_count; ++i)
+        weights.push_back((this->*weightGetterFuncs[i])());
```
This seems a little complicated. Would it not be simpler to just copy from `m_weights` instead?

lemirep replied:
Good point

DaveeFTW commented on src/core/morphcontroller.cpp:
```diff
@@ -0,0 +1,393 @@
+
+void MorphController::setWeight0(float weight0)
+{
+    if (weight0 == m_weights[0])
```
Use a fuzzy comparison here

lemirep replied:
I don't think we want to do that, we want to compare the exact values rather than whether the two are close values.

jjcasmar commented on src/core/gltf2importer/meshparser.cpp:
```diff
@@ -219,6 +247,74 @@ bool MeshParser::parse(const QJsonArray &meshArray, GLTF2Context *context)
             primitive.hasColorAttr = hasColorAttr;
         }
 
+        // All primitives in a mesh are required to declare the same number of
```
Is this glTF requirement or Kuesa?

lemirep replied:
I think it's an implicit requirement from glTF (Spec says: All primitives are required to list the same number of targets in the same order)

jjcasmar commented on src/core/metallicroughnessmaterial.cpp:
```diff
@@ -322,6 +323,15 @@ WrappedSignal<OutputType> wrapParameterSignal(MetallicRoughnessMaterial *self, S
+    \property morphController
+
+    The morph controller instance used to control morph targets weights.
+    This can be null if a material is not used without morph targets.
```
"a material is not used without morph targets" means a material that uses morph targets?

lemirep replied:
Fixed, should be "with" and not "without".

jjcasmar commented on src/core/shaders/es2/kuesa_morphtargets_normal.inc.vert:
```diff
@@ -0,0 +1,46 @@
+
+vec3 kuesa_morphNormal(const in vec3 vNormal)
+{
+    vec3 norm = vNormal;
```
if the primitive only contains 2 morphtargets, isnt this doing unneded computations?

lemirep replied:
yes but handling it this way is I think for now the best compromise between performance/maintenance until we find a way to provide multiple shader node tree description to the shader builder in a sane way.

jjcasmar commented on src/core/gltf2importer/animationparser.cpp:
```diff
@@ -361,8 +383,8 @@ Qt3DAnimation::QChannel AnimationParser::animationChannelDataFromBuffer(const QS
 {
-    const QJsonValue &inputValue = samplerObject[KEY_INPUT];
-    const QJsonValue &outputValue = samplerObject[KEY_OUTPUT];
+    const QJsonValue inputValue = samplerObject.value(KEY_INPUT);
```
probably for future patches, we should avoid modifiying this kind of stuff. It add noise to the feature we are trying to review

lemirep replied:
well it's done as a dedicated commit https://github.com/KDAB/kuesa/pull/79/commits/ae8c202138e5412c6524dec2267c881bd44fc494
it's just that if you review with the all files changes, you don't see that.

jjcasmar commented on src/core/gltf2importer/gltf2parser.cpp:
```diff
@@ -864,14 +868,50 @@ void GLTF2Parser::generateTreeNodeContent()
+                        const quint8 morphTargetCount = std::min(meshData.morphTargetCount, quint8(8));
+
+                        if (meshData.morphTargetCount > 8)
+                            qCWarning(kuesa) << "Kuesa only supports up to 8 morph targets per mesh";
```
does it make sense to continue if we are not going to animate all the morph targets?

lemirep replied:
I think so, at least you'll be able to have something working and the warning, rather than the warning and nothing

jjcasmar commented on src/core/gltf2importer/meshparser_utils.cpp:
```diff
@@ -75,13 +86,28 @@ QVarLengthArray<Qt3DRender::QAttribute::VertexBaseType, 3> validVertexBaseTypesF
         return { 3 };
     if (attributeName == Qt3DRender::QAttribute::defaultTangentAttributeName())
-        return { 4 };
+        return { 3, 4 };
```
why can the tangent have 3 components?

lemirep replied:
well because the the morph target sample files are only providing 3

jjcasmar replied:
Okay, so I guess is just setting the values of the tangents and the base configuration (the fourth component) is unchanged.

jjcasmar commented on src/core/metallicroughnessmaterial.cpp:
```diff
@@ -971,6 +999,14 @@ void MetallicRoughnessMaterial::setToneMappingAlgorithm(MetallicRoughnessEffect:
+    if (morphController == m_morphControllerParameter->value().value<MorphController *>())
+        return;
+
+    m_morphControllerParameter->setValue(QVariant::fromValue(morphController));
```
Im not sure I get this. You are setting a pointer to a custom class for the parameter, which is the uniform in the shader?

lemirep replied:
MorphController is a ShaderData, setting a ShaderData on a QParameter is possible and is used when you need to set a struct for a uniform

jjcasmar commented:
What happens if we have a mesh with a position morph target and without normals or tangents. In this case we need to generate the normals or tangents for the primitive, but how do we also need to generate a morph target for normals/tangents?

jjcasmar commented:
I have added an issue in the gltf2 specification github to try to answer the normal/tangent generation. Meanwhile, we can go as it is

jjcasmar commented on src/core/shaders/graphs/metallicroughness.frag.json:
```diff
@@ -2117,15 +2117,15 @@
                         "minor": 0
                     },
-                    "substitution": "highp $type $output; if ($condition) discard; $output = $value;"
+                    "substitution": "highp $type $output = $value; if ($condition) discard;"
```
With your last changes on the shader generator, isnt it more efficient to leave the assignment (which is going to perform the computation) after checking if we need to discard?

lemirep replied:
afaik I had to change as the shader generator expect something like * var = assignment; *
So had to change the expression to make it work though I suspect this might result in the discard part completely being removed. Which would explain why the opacity is not working properly recently

lemirep replied:
Fixed the discard if. Any substituion that requires more than one line or statement (more than one ;) should call a function. The qshadergenerator might not read it correctly or optimize it away when it shouldn't.

jjcasmar replied:
nice. probably should be documented on the node editor and shader builder in qtbase.

mkrus commented on src/core/shaders/es2/kuesa_morphtargets_normal.inc.vert:
```diff
@@ -0,0 +1,42 @@
+// Note: it is assumed the morphWeights structure uniform has been defined
+// elsewhere and included prior to this
+
+vec3 kuesa_morphNormal(const in vec3 vNormal)
```
doesn't this need precision qualifiers?  (same with other shaders)

lemirep replied:
shouldn't be needed, kuesa_morphtargets_weights.inc.frag defines precision highp float;
precision highp vec3; and is always supposed to be included before the other morphtarget includes
